File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.py
from sys import stdin
import heapq

n = int(stdin.readline())
m = int(stdin.readline())
heap = []
parents = [i for i in range(n+1)]

def find_parents(a):
    while a != parents[a]:
        a = parents[a]
    return a

def union_parents(a, b):
    a = find_parents(a)
    b = find_parents(b)
    if a > b:
        parents[a] = b
    else:
        parents[b] = a 

for i in range(m):
    a, b, c = map(int, stdin.readline().split())
    heapq.heappush(heap, (c, a, b))
        
result = 0
while heap:
    c, a, b = heapq.heappop(heap)
    if find_parents(a) != find_parents(b):
        union_parents(a, b)
        result += c

print(result)

File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] parents;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        parents = new int[n + 1];
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        for (int i = 0; i < n + 1; i++) {
            parents[i] = i;
        }

        StringTokenizer st;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken()), b = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());
            heap.offer(new int[]{c, a, b});
        }

        int result = 0;
        while (!heap.isEmpty()) {
            int[] cab = heap.poll();
            if (findParents(cab[1]) != findParents(cab[2])) {
                unionParents(cab[1], cab[2]);
                result += cab[0];
            }
        }

        System.out.println(result);
    }

    public static int findParents(int a) {
        while (a != parents[a]) {
            a = parents[a];
        }
        return a;
    }

    public static void unionParents(int a, int b) {
        a = findParents(a);
        b = findParents(b);
        if (a > b) {
            parents[a] = b;
        } else {
            parents[b] = a;
        }
    }
}